File: 27-system-ai/6-core-ml-algorithms/main.cpp
// Core ML Algorithms - C++ example (Simple Linear Regression)

#include <iostream>
#include <vector>

using namespace std;

class LinearRegression {
public:
    double coeff = 0;
    double intercept = 0;

    void fit(const vector<double>& X, const vector<double>& y) {
        int n = X.size();
        double sum_x = 0, sum_y = 0, sum_xx = 0, sum_xy = 0;
        for (int i = 0; i < n; ++i) {
            sum_x += X[i];
            sum_y += y[i];
            sum_xx += X[i] * X[i];
            sum_xy += X[i] * y[i];
        }
        coeff = (n * sum_xy - sum_x * sum_y) / (n * sum_xx - sum_x * sum_x);
        intercept = (sum_y - coeff * sum_x) / n;
    }

    double predict(double x) {
        return coeff * x + intercept;
    }
};

int main() {
    vector<double> X = {1, 2, 3, 4, 5};
    vector<double> y = {3, 6, 9, 12, 15};

    LinearRegression model;
    model.fit(X, y);

    cout << "Predictions: " << model.predict(6) << " " << model.predict(7) << endl;
    return 0;
}

File: 27-system-ai/6-core-ml-algorithms/main.java
// Core ML Algorithms - Java example (Simple Linear Regression)

class Main {
    static class LinearRegression {
        double coeff = 0;
        double intercept = 0;

        public void fit(double[] X, double[] y) {
            int n = X.length;
            double sumX = 0, sumY = 0, sumXX = 0, sumXY = 0;
            for (int i = 0; i < n; i++) {
                sumX += X[i];
                sumY += y[i];
                sumXX += X[i] * X[i];
                sumXY += X[i] * y[i];
            }
            coeff = (n * sumXY - sumX * sumY) / (n * sumXX - sumX * sumX);
            intercept = (sumY - coeff * sumX) / n;
        }

        public double predict(double x) {
            return coeff * x + intercept;
        }
    }

    public static void main(String[] args) {
        double[] X = {1, 2, 3, 4, 5};
        double[] y = {3, 6, 9, 12, 15};

        LinearRegression model = new LinearRegression();
        model.fit(X, y);

        System.out.println("Predictions: " + model.predict(6) + " " + model.predict(7));
    }
}
